### PR TITLE
Annotations improvements

### LINKED ISSUE #4258: [BUG] Relative positioning of annotations tuple cannot mix integer and float
## Details

Change the MWE in #3634 such that the relative position is `(1, 0.5)` or `(0.5, 1)` leads to error like 

```
Failed to show value:

MethodError: no method matching locate_annotation(::Plots.Subplot{Plots.GRBackend}, ::Tuple{Float64, Int64}, ::Plots.PlotText)
```

### Backends

This bug occurs on ( insert `x` below )

Backend      | yes | no  | untested
-------------|-----|-----|---------
gr (default) |   x  |     |
pyplot       |     |     |x
plotlyjs     |     |     |x
pgfplotsx    |     |     |x
unicodeplots |     |     |x
inspectdr    |     |     |x
gaston       |     |     |x

### Versions

Plots.jl version: v1.31.1
Backend  version (`]st -m <backend(s)>`):  GR v0.64.4
Output of `versioninfo()`:
```
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6850K CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, broadwell)
```

### LINKED ISSUE #4708: [FR] `annotate!` for 3D plots
`annotate!` doesn't appear to work, at least for the `gr()` backend. `annotate!` works in 2D, but for it gives warnings and produces no annotations.
```julia
julia> xyz = ntuple(_ -> randn(4), 3); scatter(xyz...); annotate!(xyz..., string.(1:4))
┌ Warning: Unused font arg: ["1", "2", "3", "4"] (Vector{String})
└ @ Plots ~/.julia/packages/Plots/9Q9pN/src/components.jl:280
┌ Warning: Unused font arg: ["1", "2", "3", "4"] (Vector{String})
└ @ Plots ~/.julia/packages/Plots/9Q9pN/src/components.jl:280
┌ Warning: Unused font arg: ["1", "2", "3", "4"] (Vector{String})
└ @ Plots ~/.julia/packages/Plots/9Q9pN/src/components.jl:280
┌ Warning: Unused font arg: ["1", "2", "3", "4"] (Vector{String})
└ @ Plots ~/.julia/packages/Plots/9Q9pN/src/components.jl:280
```

### PR DESCRIPTION
**Pull request by ivan-boikov**
## Description

- type inference and reduced code size in `process_annotation` and `locate_annotation`
- allow for mixed type in coordinate ```tuple```s (will close #4258)
- fixes problem noted in #2591 about PGFPlotsX crash
- GR: support for 3D annotations (would close #4708, at least for GR)
- PGFPlotsX: relative placement in 2D and support for 3D annotations 
- PythonPlot: support for 3D annotations (rotation is ignored, an issue with upstream?)
- InspectDR: relative placement in 2D
- PlotlyJS: 3D annotations still don't work as intended, but at least do not crash, I also don't know how to fix rotation

<details>
<summary>Figures</summary>

<details>
<summary>Generating code</summary>

```julia
using Plots
using Random

function test_2d(; N=8)
    Random.seed!(0)
    xy = ntuple(_ -> randn(N), 2); 
    pl0 = plot(xy...; marker=:circle, legend=:none, size=(2,2).*120); 
    annotate!(0.0, 0.0, text("0", 16, :red));
    annotate!((0.3, 0.3), text("REL", 12, :green, rotation=45));
    annotate!((5//6, 0.5), text("REL mixed", 12, :green, rotation=45));
    annotate!(xy..., @. text(string(1:N), :left));
    annotate!(:topcenter, "TOP");
    annotate!([:bottomcenter], ["BOT"]);
    annotate!([:topleft, :bottomright], ["TL", :BR])
end 

function test_3d(; N=8)
    Random.seed!(0)
    xyz = ntuple(_ -> randn(N), 3); 
    plot(xyz...; marker=:circle, legend=:none, size=(2,2).*120); 
    annotate!(0.0, 0.0, 0.0, text("0", :red));
    annotate!((0.3, 0.3, 0.3), text("REL", :green, rotation=45));
    annotate!((1, 0.5, 1//2), text("REL mixed", :green, rotation=45));
    annotate!(xyz..., @. text(string(1:N), :left))
end

gr()
savefig(test_2d(), "gr-2d.png")
savefig(test_3d(), "gr-3d.png")

pgfplotsx()
PGFPlotsX.latexengine!(PGFPlotsX.PDFLATEX)
savefig(test_2d(), "pgfx-2d.png")
savefig(test_3d(), "pgfx-3d.png")

inspectdr()
savefig(test_2d(), "idr-2d.png")
# no 3D

pythonplot()
savefig(test_2d(), "pyplot-2d.png")
savefig(test_3d(), "pyplot-3d.png")

plotlyjs()
savefig(test_2d(), "plotlyjs-2d.png")
savefig(test_3d(), "plotlyjs-3d.png")
```
</details>

In all cases master crashes, figures show PR

- GR
gr-2d gr-3d

- PGFPlotsX
pgfx-2d pgfx-3d

- PythonPlot
pyplot-2d pyplot-3d

- InspectDR (2D only)
idr-2d

- Plotly
plotlyjs-2d plotlyjs-3d

</details>

## Attribution
- [ ] I am listed in [.zenodo.json](https://github.com/JuliaPlots/Plots.jl/blob/2463eb9f8065c52ed8314f6e541664c5b9db88d2/.zenodo.json) (see https://github.com/JuliaPlots/Plots.jl/issues/3503)

### CODE REVIEW

**Comment by BeastyBlacksmith** [comment]
> Interestingly, tests fail only on ubuntu, while on my linux machine they pass. Marked as draft.

Its the GraphRecipes tests that fail, not Plots.

**Comment by ivan-boikov** [comment]
I think I found the issue. If I understood correctly, in GraphPlots annotations are constructed in form `(x, y, label, fontsize)` here https://github.com/JuliaPlots/GraphRecipes.jl/blob/d11908e87e425a7c4b9da2ffd7fd709097f9134e/src/graphs.jl#L1137 which I don't think is against the documentation https://docs.juliaplots.org/stable/api/#Plots.annotate!-Tuple, it should be (x, y, text(label, fontsize)). Moreover, a `label` passed can be a number too (that was the case in a failed test). So `process_annotation` can get something like `(1.2, 3.4, 1, 10)` and god knows if the 3rd number is a `z` or a `label` and PR's `process_annotation` got confused.
Maybe the best approach would be to internally convert annotations to named tuples or structs to prevent confusion once and for all. I am not sure I can do this cleanly without breaking changes (who knows, maybe there are other recipes doing the same thing), but I'll try.
Why windows and macos tests succeeded though? There are no GraphRecipes tests on those?

**Comment by BeastyBlacksmith** [comment]
> it should be `(x, y, text(label, fontsize))`

Well, you can't do that in a recipe since `text` is part of Plots and not RecipesBase. 

> Why windows and macos tests succeeded though?

Yes, upstream tests are only run on ubuntu.

> Before if lengths of annotation coordinates and annotation strings mismatched

Unfortunately we can't change that in a minor release, but one reason more for explicit cycling, that I mentioned in https://github.com/JuliaPlots/Plots.jl/issues/4565

**Comment by ivan-boikov** [comment]
I revised the processing, it has cycling and is still compact. Also tests are passed now, so marking as ready.

File: src/components.jl [hunk 1]
```diff
@@ -648,75 +657,41 @@ function _relative_position(xmin, xmax, pos::Length{:pct}, scale::Symbol)
...
-function locate_annotation(sp::Subplot, pos::Symbol, label::PlotText)
-    x, y = position_multiplier[pos]
-    (
-        _relative_position(
-            axis_limits(sp, :x)...,
-            x,
-            sp[get_attr_symbol(:x, :axis)][:scale],
-        ),
+locate_annotation(sp::Subplot, rel::Tuple, label::PlotText) = (
+    map(1:length(rel), [:x, :y, :z]) do i, letter
```
Comment (t-bltg) [inline src/components.jl:680]:
```suggestion
    map(1:length(rel), (:x, :y, :z)) do i, letter
```

Use `Tuple`s where you can (statically sized).

**Review by t-bltg — APPROVED** [review]
LGTM

File: src/components.jl (hunk 1 above)
Comment (ivan-boikov) [reply src/components.jl]: Good point!

**Comment by BeastyBlacksmith** [comment]
Good job! Thank you 🙂

**Pull request merged**